### ISSUE TITLE
Shutdown Logging Support

### ISSUE THREAD

**Issue by meltsufin** [GoogleCloudPlatform/openjdk-runtime#114]
**TL;DR: On a VM shutdown, allow logging the application’s thread dump and heap info to Cloud Logging.**

On App Engine, sometimes autohealer decides to kill unhealthy VMs. In that case we don't have sufficient logging done from the VM to debug if it was an issue with the app or the VM. This can happen when the app code gets stuck (like deadlock) and stopped returning requests, including health checks. 

Because it is possible that the JVM is ailing, it might not have enough resources to log to Cloud Logging directly. So we will be using stdout + fluentd to capture app container logs.

Explicit user intent is required for potentially sensitive information. The following environment variables will be used to enable/disable app container shutdown reporting (must be set to true or false):
 
`SHUTDOWN_LOGGING_THREAD_DUMP` - save thread traces
`SHUTDOWN_LOGGING_HEAP_INFO` - save heap information
 
The runtime will capture `SIGTERM`, and run debugging tools on the JVM to emit the thread traces and heap information to stdout.
 
Further, for customers with a small number of VMs we default to reporting shutdown logs for all VMs. For larger customers running hundreds or thousands of instances, we would allow them to be able to sample logs. This will be done by using the environment variable `SHUTDOWN_LOGGING_SAMPLE_THRESHOLD` which is an int between 0 and 100. 0 means no VMs report logs, 100 means all VMs report logs. (If this env var is not set, we default to reporting for all VMs). The runtime will then generate a random number between 0 (inclusive) and 100 (exclusive) and if it’s less than the threshold set, it should do the trap-and-log logic described above.

@karan
